packages/beluga/src/allocator.rs

```diff
@@ -1,81 +1,111 @@
+use core::alloc::Layout;
 use core::ffi::c_void;
 use core::fmt::Debug;
+use core::mem::{align_of, size_of};
 
-use aws_c_common_sys::{aws_allocator, aws_default_allocator};
+use aws_c_common_sys::aws_allocator;
 
-pub type AllocatorRef = &'static Allocator;
-
-impl Default for AllocatorRef {
-    fn default() -> Self {
-        Allocator::default()
-    }
-}
+static GLOBAL_ALLOCATOR: GlobalAllocator = GlobalAllocator::new();
 
 #[repr(transparent)]
-pub struct Allocator(aws_allocator);
+pub struct GlobalAllocator(aws_allocator);
 
-impl Allocator {
-    /// # Safety
-    ///
-    /// `allocator` must be a valid (non-null) pointer with static lifetime.
+impl GlobalAllocator {
     #[inline]
-    pub(crate) unsafe fn new(allocator: *mut aws_allocator) -> AllocatorRef {
-        debug_assert!(!allocator.is_null());
-        &mut *allocator.cast()
+    const fn new() -> Self {
+        Self(aws_allocator {
+            mem_acquire: Some(mem_acquire),
+            mem_release: Some(mem_release),
+            mem_realloc: Some(mem_realloc),
+            mem_calloc: Some(mem_calloc),
+            impl_: core::ptr::null_mut(),
+        })
     }
 
     #[inline]
     #[must_use]
-    pub fn default() -> AllocatorRef {
-        unsafe { Self::new(aws_default_allocator()) }
+    pub fn ptr() -> *mut aws_allocator {
+        let allocator = &GLOBAL_ALLOCATOR;
+        (&allocator.0 as *const aws_allocator).cast_mut()
     }
+}
 
-    #[inline]
-    #[must_use]
-    pub const fn as_ptr(&self) -> *mut aws_allocator {
-        core::ptr::addr_of!(self.0).cast_mut()
-    }
+/// Allocates memory for the given size using the standard memory allocator.
+///
+/// # Arguments
+/// - `size` - size of desired allocation. Assume that size is properly aligned.
+///
+/// # Safety
+///
+/// The caller must ensure the given size is valid. The returned pointer
+/// must be properly deallocated later.
+#[inline]
+unsafe extern "C" fn mem_acquire(_: *mut aws_allocator, size: usize) -> *mut c_void {
+    // Calculates the Layout to allocate for the block, including space for
+    // storing the Layout at the start.
+    let layout =
+        Layout::from_size_align_unchecked(size + size_of::<Layout>(), align_of::<Layout>());
+    // Allocates memory using the provided layout, writes the layout to the
+    // start of the allocated block, and returns a pointer offset past the
+    // written layout for use as the allocated memory region.
+    let ptr = std::alloc::alloc(layout);
```

Comment:
we should check if ptr is null (this is unfortunately allowed with the `GlobalAlloc` trait). If null, we can just return null to aws, which will handle this case correctly.